op-program/client/interop/consolidate.go

```diff
@@ -0,0 +1,227 @@
+)
+
+func ReceiptsToExecutingMessages(depset depset.ChainIndexFromID, receipts ethtypes.Receipts) ([]*supervisortypes.ExecutingMessage, uint32, error) {
+	var execMsgs []*supervisortypes.ExecutingMessage
+	var logCount uint32
+	for _, rcpt := range receipts {
+		logCount += uint32(len(rcpt.Logs))
+		for _, l := range rcpt.Logs {
+			execMsg, err := processors.DecodeExecutingMessageLog(l, depset)
+			if err != nil {
+				return nil, 0, err
+			}
+			// TODO: e2e test for both executing and non-executing messages in the logs
+			if execMsg != nil {
+				execMsgs = append(execMsgs, execMsg)
+			}
+		}
+	}
+	return execMsgs, logCount, nil
+}
+
+func RunConsolidation(deps ConsolidateCheckDeps,
+	oracle l2.Oracle,
+	transitionState *types.TransitionState,
+	superRoot *eth.SuperV1,
+) (eth.Bytes32, error) {
+	var consolidatedChains []eth.ChainIDAndOutput
+
+	for i, chain := range superRoot.Chains {
+		progress := transitionState.PendingProgress[i]
+
+		// TODO(13776): hint block data execution in case the pending progress is not canonical so we can fetch the correct receipts
+		block, receipts := oracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
+		execMsgs, _, err := ReceiptsToExecutingMessages(deps.DependencySet(), receipts)
+		if err != nil {
+			return eth.Bytes32{}, err
+		}
+
+		candidate := supervisortypes.BlockSeal{
+			Hash:      progress.BlockHash,
+			Number:    block.NumberU64(),
+			Timestamp: block.Time(),
+		}
+		hazards, err := getHazards(progress.BlockHash, execMsgs)
+		if err != nil {
+			return eth.Bytes32{}, err
+		}
+		if err := checkHazards(deps, &candidate, hazards); err != nil {
+			// TODO(13776): replace with deposit-only block if ErrConflict, ErrCycle, or ErrFuture
+			return eth.Bytes32{}, err
+		}
+		consolidatedChains = append(consolidatedChains, eth.ChainIDAndOutput{
+			ChainID: chain.ChainID,
+			// TODO(13776): when applicable, use the deposit-only block output root
+			Output: progress.OutputRoot,
+		})
+	}
+	consolidatedSuper := &eth.SuperV1{
+		Timestamp: superRoot.Timestamp + 1,
+		Chains:    consolidatedChains,
+	}
+	return eth.SuperRoot(consolidatedSuper), nil
+}
+
+type ConsolidateCheckDeps interface {
+	cross.UnsafeFrontierCheckDeps
+	cross.CycleCheckDeps
+}
+
+func checkHazards(
+	deps ConsolidateCheckDeps,
+	candidate *supervisortypes.BlockSeal,
+	hazards map[supervisortypes.ChainIndex]supervisortypes.BlockSeal,
+) error {
+	// TODO: for each of these hazard checks, ensure that they're scoped to the super root timestamp of the chain
+	if err := cross.HazardUnsafeFrontierChecks(deps, hazards); err != nil {
+		return err
+	}
+	if err := cross.HazardCycleChecks(deps.DependencySet(), deps, candidate.Timestamp, hazards); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getHazards treats all executing messages as hazards
+func getHazards(
+	blockHash common.Hash,
+	execMsgs []*supervisortypes.ExecutingMessage,
+) (map[supervisortypes.ChainIndex]supervisortypes.BlockSeal, error) {
```

Comment:
Hazards in the supervisor are only those that are in the same timestamp, and thus need cycle detection. Hence the "frontier" name on the function.

The other executing messages can be checked for correct inclusion, without cycle detection work.

By assuming everything as "hazard", the cycle detection has to do more work than needed, and I am not sure if it handles these older messages correctly, since it was built to just do the same-timestamp checks.

Reply:
Understood. I misread Frontier checks as also checking for correct inclusion. Turns out the `CrossSafeHazards` check filters executing messages for the current timestamp while also checking if they're executable.

Fixed.